pkg/controllers/routereflector/ctrl_ops.go

```diff
@@ -0,0 +1,243 @@
+			log.Errorf("Failed to revert update node %s because of %s", calicoNode.GetName(), err.Error())
+			return err
+		}
+
+		delete(c.routeReflectorsUnderOperation, calicoNode.GetUID())
+
+		return c.update(calicoNode)
+	}
+
+	return nil
+}
+
+func (c *ctrl) updateNodeLabels(nodeLoader loaders.NodeLoader) error {
+	missingRouteReflectors := 0
+
+	kubeNodes, err := nodeLoader.List()
+	if err != nil {
+		log.Errorf("Unable to fetch kube nodes because of %s", err.Error())
+		return err
+	}
+
+	affectedNodes := c.collectAffectedNodesAndStatuses(kubeNodes)
+
+	for _, status := range c.Topology.GetRouteReflectorStatuses(affectedNodes) {
+		status.ExpectedRRs += missingRouteReflectors
+		log.Infof("Route refletor status in zone(s) %v (actual/expected) = %d/%d", status.Zones, status.ActualRRs, status.ExpectedRRs)
+
+		for _, n := range status.Nodes {
+			n := c.nodes[n.GetUID()]
+
+			if affectedNodes[n] {
+				continue
+			} else if status.ExpectedRRs == status.ActualRRs {
+				break
+			}
+
+			if diff := status.ExpectedRRs - status.ActualRRs; diff != 0 {
+				if updated, err := c.updateRRStatus(n, nodeLoader, diff); err != nil {
+					log.Errorf("Unable to update node %s because of %s", n.GetName(), err.Error())
+					return err
+				} else if updated && diff > 0 {
+					status.ActualRRs++
+				} else if updated && diff < 0 {
+					status.ActualRRs--
+				}
+			}
+		}
+
+		if status.ExpectedRRs != status.ActualRRs {
+			log.Infof("Actual number %d is different than expected %d", status.ActualRRs, status.ExpectedRRs)
+			missingRouteReflectors = status.ExpectedRRs - status.ActualRRs
+		}
+	}
+
+	if missingRouteReflectors != 0 {
+		log.Errorf("Actual number is different than expected, missing: %d", missingRouteReflectors)
+	}
+
+	return nil
+}
+
+func (c *ctrl) updateBGPTopology(calicoNode *apiv3.Node, nodeLoader loaders.NodeLoader) error {
+	existingBGPPeers, err := c.BGPPeer.list()
+	if err != nil {
+		log.Errorf("Unable to list BGP peers because of %s", err.Error())
+		return err
+	}
+
+	log.Debugf("Existing BGPeers are: %v", existingBGPPeers.Items)
+
+	kubeNodes, err := nodeLoader.List()
+	if err != nil {
+		log.Errorf("Unable to fetch kube nodes because of %s", err.Error())
+		return err
+	}
+
+	affectedNodes := c.collectAffectedNodesAndStatuses(kubeNodes)
+
+	toRefresh, toRemove := c.Topology.GenerateBGPPeers(c.nodes, affectedNodes, existingBGPPeers)
+
+	log.Infof("Number of BGPPeers to refresh: %v", len(toRefresh))
+	log.Debugf("To refresh BGPeers are: %v", toRefresh)
+	log.Infof("Number of BGPPeers to delete: %v", len(toRemove))
+	log.Debugf("To delete BGPeers are: %v", toRemove)
+
+	for _, bp := range toRefresh {
+		log.Infof("Saving %s BGPPeer", bp.Name)
+		if err := c.BGPPeer.save(&bp); err != nil {
+			log.Errorf("Unable to save BGPPeer because of %s", err.Error())
+			return err
+		}
+	}
+
+	// Give some time to Calico to establish new connections before deleting old ones
+	if len(toRefresh) > 0 {
+		time.Sleep(2)
```

Comment:
This might be a question that how to avoid the situation when a node looses it's all route reflectors. In the current implementation delete doesn't happen until all active route reflectors are all configured. This sleep gives some time to Calico to establish new BGP sessions.